third_party/move/tools/move-resource-viewer/src/lib.rs

```diff
@@ -231,18 +230,50 @@ impl<V: CompiledModuleView> MoveValueAnnotator<V> {
             args.len(),
         );
 
-        // Make an approximation at the fat types for the type arguments
-        let ty_args: Vec<FatType> = ty_args.iter().map(|inner| inner.into()).collect();
-
-        types
+        // Convert type args
+        let ty_args: Vec<FatType> = ty_args
             .iter()
-            .enumerate()
-            .map(|(i, ty)| {
-                ty.subst(&ty_args, limit)
-                    .map_err(anyhow::Error::from)
-                    .and_then(|fat_type| self.view_value_by_fat_type(&fat_type, &args[i], limit))
-            })
-            .collect::<anyhow::Result<Vec<AnnotatedMoveValue>>>()
+            .map(|ty| self.resolve_type_impl(ty, limit))
+            .collect::<anyhow::Result<_>>()?;
+
+        if ty_args.is_empty() {
+            types
+                .iter()
+                .enumerate()
+                .map(|(i, ty)| self.view_value_by_fat_type(ty, &args[i], limit))
+                .collect::<anyhow::Result<Vec<AnnotatedMoveValue>>>()
+        } else {
+            types
+                .iter()
+                .enumerate()
+                .map(|(i, ty)| {
+                    ty.subst(&ty_args, &self.struct_substitutor(), limit)
+                        .map_err(anyhow::Error::from)
+                        .and_then(|fat_type| {
+                            self.view_value_by_fat_type(&fat_type, &args[i], limit)
+                        })
+                })
+                .collect::<anyhow::Result<Vec<AnnotatedMoveValue>>>()
+        }
+    }
+
+    fn struct_substitutor(
+        &self,
+    ) -> impl Fn(&FatStructType, &[FatType], &mut Limiter) -> PartialVMResult<FatStructRef> {
+        |st, ty_args, limiter| {
+            let st_ty_args = st
```

Comment:
let's assert `ty_args` are not empty? So that users of this can clone instead + add a comment about this